tests/unit/s2n_ktls_test.c

```diff
@@ -13,13 +13,59 @@
  * permissions and limitations under the License.
  */
 
+#include "tls/s2n_ktls.h"
+
 #include "crypto/s2n_cipher.h"
+#include "error/s2n_errno.h"
+#include "s2n.h"
 #include "s2n_test.h"
+#include "stdio.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_handshake_type.h"
+#include "utils/s2n_safety.h"
+
+#define S2N_TEST_INADDR_LOOPBACK 0x7f000001 /* 127.0.0.1 */
+
+S2N_RESULT s2n_ktls_validate(struct s2n_connection *conn);
+S2N_RESULT s2n_ktls_validate_socket_mode(struct s2n_connection *conn, s2n_ktls_mode ktls_mode);
+S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, int *fd);
+S2N_RESULT s2n_ktls_configure_socket(struct s2n_connection *conn, s2n_ktls_mode ktls_mode);
+S2N_RESULT s2n_ignore_ktls_ulp_for_testing(void);
+
+S2N_RESULT s2n_test_configure_ktls_connection(struct s2n_connection *conn, int *fd)
+{
+    *fd = 1;
+    EXPECT_SUCCESS(s2n_connection_set_write_fd(conn, *fd));
+    EXPECT_SUCCESS(s2n_connection_set_read_fd(conn, *fd));
+
+    conn->managed_send_io = true;
+    conn->ktls_send_enabled = false;
+    conn->managed_recv_io = true;
+    conn->ktls_recv_enabled = false;
+
+    conn->initial->cipher_suite->record_alg = &s2n_record_alg_aes128_gcm;
+    conn->actual_protocol_version = S2N_TLS12;
+
+    return S2N_RESULT_OK;
+}
 
 int main(int argc, char **argv)
 {
     BEGIN_TEST();
 
+    EXPECT_OK(s2n_ignore_ktls_ulp_for_testing());
+
+    /* kTLS feature probe */
+    {
+#ifdef __FreeBSD__
+    #ifndef S2N_PLATFORM_SUPPORTS_KTLS
+        /* https://github.com/torvalds/linux/commit/3c4d7559159bfe1e3b94df3a657b2cda3a34e218
+        * kTLS support was first added in linux 4.13.0. */
+        FAIL_MSG("kTLS feature probe is not working");
```

Comment:
Do all versions of FreeBSD support kTLS? It seems like your previous linux test for this that checked the version would be better since it's more specific?

Reply:
https://github.com/aws/s2n-tls/pull/3808/commits/85a2fcab9935fa492ddf7d44059bad9aa584a251